### PR TITLE
feature: add Bayesian Machine Scientist

### PR DESCRIPTION
**Pull request by hollandjg**
## Description

Merge the Bayesian Machine Scientist from https://github.com/AutoResearch/Bayesian-Machine-Scientist into the AutoRA package.

To-Do:

- [ ] Check that all the tests still work
- [ ] Ensure documentation is up to date
- [ ] Bump version number to 1.1.0

### Type of change:
- New feature (non-breaking change which adds functionality)

### Features: 
- Added the Bayesian Machine Scientist theorist

### CODE REVIEW

**Comment by hollandjg** [comment]
Hi @benwandrew @TheLemonPig @listar2000 
This is looking great! There's a couple of to-dos which have appeared as a result of the merge before we can close this out
- [ ] Check and update the documentation – some of the math bits aren't quite formatted right:
Screen Shot 2022-11-11 at 15 42 55

You can run the documentation server locally to test it by:
-  installing the dependencies locally with `poetry install` and 
- then running `poetry run mkdocs serve`.

Send me a message on slack if you can't get the poetry environment working, there are some weird issues happening right now with PyTorch.

- [ ] Decide whether we need the docs/theorist/bms/Tutorial.ipynb, if we do, then keep it, if we don't then delete it.
- [ ] Delete any unused images from the `docs/theorist/bms/img` directory – I think you rewrote the BMSMath images as MathJax equations, so they can probably go. Also the BayesianMath.png looks like it's unused.

**Important**: Any changes you need to make need to be made on this PR – it's too tricky to make any additional merges from the Bayesian-Machine-Scientist repo now.

**Comment by hollandjg** [comment]
Also, please go through the code one last time and make sure there's nothing which you wouldn't want to be available on the internet for ever – anything unprofessional, passwords, curse words... You name it. Once we merge, this will be completely public and your future employers may well look at it! Once you've checked it over, please tick here, and make a note that you've reviewed everything in your code review.
- [ ] Josh
- [ ] Ben
- [ ] Star

**Comment by TheLemonPig** [comment]
The fact that the server updates local changes in real time is so convenient :)

**Review by listar2000 — APPROVED** [review]
I mainly took a look at the documentation parts (e.g. the math formulation & image included) and they LGTM! Great job @TheLemonPig

**Review by TheLemonPig — APPROVED** [review]
Thanks @listar2000! :)

**Review by benwandrew — APPROVED** [review]
looks like a clean, accretive merger to me!

**Comment by benwandrew** [comment]
@TheLemonPig I made some changes and caught a few more issues with the documentation. Can you please review my latest commits? Thanks!

File: docs/theorist/bms/how_it_works.md [hunk 1]
```diff
@@ -38,19 +38,19 @@ $\mathscr{L}(f_i)\approx\dfrac{B(f_i)}{2} - \log[P(f_i)]$
 
 where $B(f_i)=k\log[n] - 2\log[P(D|\theta^*,f_i)]$
 
-In this formulation, the goodness of fit $p(D|\theta^*,f_i)$ and likelihood $p(f_i)$ of an equation are equally and logarithmically weighted to each other - e.g., improving the fit by a factor of 2 is offset by having half the likelihood.
+In this formulation, the goodness of fit $p(D|\theta^*,f_i)$ and likelihood $p(f_i)$ of an equation are equally and logarithmically weighted to each other (e.g., improving the fit by a factor of 2 is offset by halving the likelihood).
 
-To better understand the problem, equations are framed in terms of acyclic graphs (i.e., trees)
+To better frame the problem, equations are cast as acyclic graphs (i.e., trees).
```
Comment (TheLemonPig) [inline docs/theorist/bms/how_it_works.md:43]: I would probably rephrase this as something like "equations are modeled as". Casting is a specific term using by coders. Using it informally here might be confusing.

**Review by TheLemonPig — APPROVED** [review]
Otherwise everything else looks good

**Review by benwandrew — APPROVED** [review]

**Comment by benwandrew** [comment]
@hollandjg is [this](https://github.com/AutoResearch/autora/pull/129/commits/1b80ef556b39d2a25b2f39ca7f7bc503acc03422) all we need for the last to-do in your PR

**Comment by hollandjg** [comment]
> @hollandjg is [this](https://github.com/AutoResearch/autora/pull/129/commits/1b80ef556b39d2a25b2f39ca7f7bc503acc03422) all we need for the bump version to-do in your PR?

Yes. I'll update the poetry.lock and the pyproject.toml so we can merge.

**Review by hollandjg — COMMENTED** [review]
Great! Let's merge it.
There are some issues we want to get to later, if this method gets more traction and proves to be useful. They are:
- Refactor, updating type annotations: #136
- Work out why it BMS throws warnings about invalid values when running some regressions #137
